language-plutus-core/test/Language/PlutusCore/Gen/Type.hs

```diff
@@ -0,0 +1,262 @@
+{-|
+Description: PLC Syntax, typechecker,semantics property based testing.
+
+This file contains
+1. A duplicate of the Plutus Core Abstract Syntax (types and terms)
+2. A kind checker and a type checker
+3. Reduction semantics for types
+-}
+
+{-# LANGUAGE DeriveDataTypeable        #-}
+{-# LANGUAGE FlexibleInstances         #-}
+{-# LANGUAGE LambdaCase                #-}
+{-# LANGUAGE RecordWildCards           #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+{-# LANGUAGE TemplateHaskell           #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Language.PlutusCore.Gen.Type
+  ( TypeBuiltinG (..)
+  , TypeG (..)
+  , ClosedTypeG
+  , toClosedType
+  , checkClosedTypeG
+  , normalizeTypeG
+  , GenError (..)
+  , ErrorP (..)
+  ) where
+
+import           Control.Enumerable
+import           Control.Monad.Except
+import           Data.Coolean
+import qualified Data.Stream                    as Stream
+import qualified Data.Text                      as Text
+import           Language.PlutusCore
+import           Language.PlutusCore.Gen.Common
+
+-- * Enumerating builtin types
+
+data TypeBuiltinG
+  = TyByteStringG
+  | TyIntegerG
+  | TyStringG
+  deriving (Typeable, Eq, Show)
+
+deriveEnumerable ''TypeBuiltinG
+
+-- |Convert generated builtin types to Plutus builtin types.
+toTypeBuiltin :: TypeBuiltinG -> Some (TypeIn DefaultUni)
+toTypeBuiltin TyByteStringG = Some (TypeIn DefaultUniByteString)
+toTypeBuiltin TyIntegerG    = Some (TypeIn DefaultUniInteger)
+toTypeBuiltin TyStringG     = Some (TypeIn DefaultUniString)
+
+
+-- * Enumerating types
+
+data TypeG n
+  = TyVarG n
+  | TyFunG (TypeG n) (TypeG n)
+  | TyIFixG (TypeG n) (Kind ()) (TypeG n)
+  | TyForallG (Kind ()) (TypeG (S n))
+  | TyBuiltinG TypeBuiltinG
+  | TyLamG (TypeG (S n))
+  | TyAppG (TypeG n) (TypeG n) (Kind ())
```

Comment:
In the regular AST we annotate lambdas with kinds, rather than applications. It would be nice to have a description of why we do things differently here.

Reply:
The reason is eagerness and efficiency. If we have the kind information at the application site, we can check the two subterms in parallel.

Reply:
Thanks.

@jmchapman could you inline @wenkokke's comment

> The reason is eagerness and efficiency. If we have the kind information at the application site, we can check the two subterms in parallel, while evaluating as little as possible.

in the code?